### ISSUE TITLE
Handling of new levels in randomForest (and other algorithms using rpart)

### ISSUE THREAD

**Issue by Prometheus77** [mlr-org/mlr#2051]
The current mlr handling of new levels in data that weren't available at time of training is suboptimal. The current approach of using fix.factors.prediction = TRUE keeps the learner from crashing but ends up producing an NA for any observation where new levels are present in new data.

An alternative method of handling the new levels problem is to create a dummy level for each factor in the training data, and then before predicting with new data, check for any levels not present at training time and set them to the dummy level. This requires training with an empty level (which mlr doesn't currently permit) but doesn't cause any problems in the actual training and prediction process. Furthermore, real-world data often contains new levels, and it is often useful to ignore that level and use all the other features to make a prediction, rather than simply returning NA.

I've provided some example code below that illustrates the point

```r
set.seed(0)

# create a data frame for training
df_train <- data.frame(x1 = sample(LETTERS[1:10], 100, replace = TRUE), 
                       x2 = rnorm(100),
                       x3 = rnorm(100),
                       y = sample(0:1, 100, replace = TRUE))

# train a randomForest on the data
task <- makeClassifTask(data = df_train, target = "y")
lrn <- makeLearner("classif.randomForest", fix.factors.prediction = TRUE)
mdl <- train(lrn, task)

# create a new set of data that we didn't have access to at training time
df_new <- data.frame(x1 = sample(LETTERS[1:11], 100, replace = TRUE), 
                     x2 = rnorm(100),
                     x3 = rnorm(100))

# run prediction on the new data
prd <- predict(mdl, newdata = df_new)
getPredictionResponse(prd) # we get NAs wherever the new data had "K" in column x1

# alternate way of handling
df_train$x1 <- factor(df_train$x1, levels = c(levels(df_train$x1), "new.level"))
new.levs <- levels(df_new$x1)[which(!(levels(df_new$x1) %in% levels(df_train$x1)))]
df_new$x1 <- factor(df_new$x1, levels = c(levels(df_new$x1), "new.level"))
df_new$x1[which(df_new$x1 %in% new.levs)] <- "new.level"
df_new$x1 <- factor(df_new$x1, levels = levels(df_train$x1))

# It works fine in randomForest without NAs
mdl <- randomForest(x = df_train[, -4], y = as.factor(df_train[, 4]), type = "classification")
prd <- predict(mdl, newdata = df_new)
print(prd)

# But mlr won't let me implement this because of Task.R line 149-150
task <- makeClassifTask(data = df_train, target = "y", fixup.data = "no")
# Throws error: Error in (function (cn, x)  : Column 'x1' contains empty factor levels.
```

Possible solutions:
1) Easy fix: When fixup.data = "no", skip the check for empty levels. Then I can at least implement this workaround in a custom preprocessing wrapper.
2) Harder fix: Create an option similar to fix.factors.prediction that implements the above method

My current workaround is to use makeDummyFeaturesWrapper so that I don't even feed it factors in the first place, but this is suboptimal as randomForest generally does better when it can treat factors as factors.

**Comment by Prometheus77** [comment]
Now that I'm thinking about it, I could probably add a fake observation to the training set in my preprocessing wrapper that contains the "new.level" for each factor, and the mean for every numeric variable as well as the mean (numeric) or mode (factor) for the response variable.

If it's just one observation, it shouldn't skew the training too bad. But still feels pretty hacky.

**Comment by pat-s** [comment]
Are you aware of https://github.com/mlr-org/mlr/pull/1882?

Its similar to your problem just for lm* models.

**Comment by Prometheus77** [comment]
Yes, similar problem, but the proposed solution (remove offending
observations) is very different. That approach is probably okay for
benchmarking, but if using a model to make decisions in a production
environment, it would seem strongly preferable to go down the decision tree
which doesn't include the offending factors, but take advantage of all the
other features to produce a result.

Interestingly, randomForest doesn't crash if the factor level is not
present in the training fold of resampling, as long as it's present in the
task. E.g. if I do a 1/3 holdout, and feature x1 contains a factor level
"A" that is only present in the holdout set, it will still predict properly
based on factors x2 ... xN, because the "level" A still existed, even if
there were no "A" observations in the training set.

Aaron

**Comment by pat-s** [comment]
I'm not removing the observation, just the problematic factor level in the respective variable. It depends on the model characteristic how the algorithm handles this case then.

There is no "good" solution to this problem - every idea is kinda hacky. 

"At least" is RF doing a prediction at all. lm* models error before the call if factor levels do not match.

**Comment by mb706** [comment]
It is possible to do this with [mlrCPO](https://github.com/mlr-org/mlrCPO):
```R
# using the datasets of OP
library("mlrCPO")
lrn = cpoImputeConstant(const = "new.level", affect.type = "factor") %>>%
  makeLearner("classif.randomForest")
mdl = train(lrn, task)
prd = predict(mdl, newdata = df_new)
getPredictionResponse(prd)
```
output:
```
  [1] 1 1 0 0 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 0 1 0 1 1 1 0
 [38] 1 0 1 1 1 0 1 1 1 1 1 0 0 1 0 1 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1 1 1 1 1 0
 [75] 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 0 0 0 1 0 1 1 1 1 1
Levels: 0
```
However, this still treats new levels as `NA`s. It just imputes them as a new level.

**Comment by jadhosn** [comment]
I was just wondering whether adding these new levels to the intercept (in the case of linear models) which is the omitted level (aka the N-1 level that we don't include) or having a new level, maybe name it "other" that will include all these new observations?
